File: cpp/reroot.cpp
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief v1.5 cexen.
 */
struct tree
{
public:
    const size_t n;

private:
    vector<vector<pair<size_t, size_t>>> _adjs;
    size_t _root = -1;
    vector<size_t> _idxs;
    vector<pair<size_t, size_t>> _parents;

public:
    tree() : tree(0) {}
    explicit tree(const size_t n) : n(n), _adjs(n), _parents(n) {}

    /**
     * Specify e (edge index) if you need it when add_e.
     */
    void connect(const size_t u, const size_t v, const size_t e = -1)
    {
        assert(0 <= u && u < n);
        assert(0 <= v && v < n);
        _adjs.at(u).emplace_back(v, e);
        _adjs.at(v).emplace_back(u, e);
    }
    const vector<vector<pair<size_t, size_t>>> &adjs() const noexcept
    {
        return _adjs;
    }
    void sort(size_t root = 0)
    {
        assert(0 <= root && root < n);
        if (_root == root)
            return;
        _root = root;
        _idxs.clear();
        vector<size_t> q{root};
        basic_string<bool> visited(n, false);
        visited.at(root) = true;
        while (!q.empty())
        {
            const auto i = move(q.back());
            q.pop_back();
            _idxs.push_back(i);
            for (const auto &adj : _adjs.at(i))
            {
                const auto &[j, e] = adj;
                if (visited.at(j))
                {
                    _parents.at(i) = adj;
                    continue;
                }
                visited.at(j) = true;
                q.push_back(j);
            }
        }
    }
    template <
        class E, class V,
        E (*add_e)(const V &, const size_t, const size_t),
        E (*merge)(const E &, const E &),
        V (*add_v)(const E &, const size_t, const size_t),
        E (*ee)()>
    vector<V> collect_for_root(const size_t root = 0)
    {
        return collect_for_root<E, V>(
            [](const V &v, const size_t i, const size_t e) constexpr { return add_e(v, i, e); },
            [](const E &l, const E &r) constexpr { return merge(l, r); },
            [](const E &e, const size_t i, const size_t parent) constexpr { return add_v(e, i, parent); },
            []() constexpr { return ee(); },
            root);
    }
    template <
        class E, class V,
        class FEV,
        class FEEE,
        class FVE,
        class FE>
    pair<vector<V>, vector<vector<E>>> collect_for_root(const FEV add_e, const FEEE merge, const FVE add_v, const FE ee, const size_t root = 0)
    {
        sort(root);
        vector<V> ans0(n);
        vector<vector<E>> edgeses(n);
        for (auto i = _idxs.rbegin(); i != _idxs.rend(); i++)
        {
            const auto &parent = _parents.at(*i);
            vector<E> &edges = edgeses.at(*i);
            auto accum_l = ee();
            for (const auto &adj : _adjs.at(*i))
            {
                const auto &[j, e] = adj;
                if (adj == parent)
                {
                    edges.emplace_back(ee());
                }
                else
                {
                    const auto edge = add_e(ans0.at(j), j, e);
                    edges.emplace_back(edge);
                    accum_l = merge(accum_l, edge);
                }
            }
            ans0.at(*i) = add_v(accum_l, *i, parent.first);
        }
        return {ans0, edgeses};
    }
    template <
        class E, class V,
        E (*add_e)(const V &, const size_t, const size_t),
        E (*merge)(const E &, const E &),
        V (*add_v)(const E &, const size_t, const size_t),
        E (*ee)()>
    vector<V> collect_for_all(const size_t root = 0)
    {
        return collect_for_all<E, V>(
            [](const V &v, const size_t i, const size_t e) constexpr { return add_e(v, i, e); },
            [](const E &l, const E &r) constexpr { return merge(l, r); },
            [](const E &e, const size_t i, const size_t parent) constexpr { return add_v(e, i, parent); },
            []() constexpr { return ee(); },
            root);
    }
    template <
        class E, class V,
        class FEV,
        class FEEE,
        class FVE,
        class FE>
    vector<V> collect_for_all(const FEV add_e, const FEEE merge, const FVE add_v, const FE ee, const size_t root = 0)
    {
        vector<vector<E>> edgeses = collect_for_root<E, V>(add_e, merge, add_v, ee, root).second;
        vector<V> ans(n);
        vector<pair<size_t, E>> q{{root, ee()}};
        while (!q.empty())
        {
            const auto [i, edge_parent] = move(q.back());
            q.pop_back();
            const auto &adjsi = _adjs.at(i);
            const auto &parent = _parents.at(i);
            vector<E> &edges = edgeses.at(i);
            vector<E> accums_l{ee()};
            assert(adjsi.size() == edges.size());
            for (size_t k = 0; k < edges.size(); ++k)
            {
                if (adjsi.at(k) == parent)
                {
                    edges.at(k) = edge_parent;
                }
                const E &edge = edges.at(k);
                accums_l.push_back(merge(accums_l.back(), edge));
            }
            accums_l.pop_back(); // not used
            E accum_r = ee();
            assert(adjsi.size() == accums_l.size());
            assert(accums_l.size() == edges.size());
            for (size_t k = edges.size(); k-- > 0;)
            {
                const auto &adj = adjsi.at(k);
                const auto &accum_l = accums_l.at(k);
                const auto &edge = edges.at(k);
                if (adj != parent)
                {
                    const auto &[j, e] = adj;
                    // myself as a child of j
                    const auto myself = add_v(merge(accum_l, accum_r), i, j);
                    q.emplace_back(j, add_e(myself, i, e));
                }
                accum_r = merge(edge, accum_r);
            }
            ans.at(i) = add_v(accum_r, i, i);
        }
        return ans;
    }
};

// using E = int32_t;
// using V = int32_t;
// auto add_e = [](const V &v, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t e) -> E
// { throw logic_error("not_implemented"); };
// auto merge = [](const E &l, const E &r) -> E
// { throw logic_error("not_implemented"); };
// auto add_v = [](const E &e, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t parent) -> V
// { throw logic_error("not_implemented"); };
// constexpr auto e = []() -> E
// { throw logic_error("not_implemented"); };

// --------------------

/**
 * @brief https://atcoder.jp/contests/dp/tasks/dp_v
 */
void solve_dp_v()
{
    size_t N, M;
    cin >> N >> M;
    tree tree(N);
    for (size_t i = 0; i < N - 1; i++)
    {
        size_t x, y;
        cin >> x >> y;
        tree.connect(x - 1, y - 1);
    }
    using E = int64_t;
    using V = int64_t;
    const auto add_e = [&M](const V &v, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t e) -> E
    {
        return (v + 1) % M;
    };
    const auto merge = [&M](const E &l, const E &r) -> E
    {
        return l * r % M;
    };
    constexpr auto add_v = [](const E &e, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t parent) -> V
    {
        return e;
    };
    constexpr auto ee = []() -> E
    {
        return 1;
    };
    const auto ans = tree.collect_for_all<E, V>(add_e, merge, add_v, ee);
    for (size_t i = 0; i < N; i++)
        cout << ans.at(i) << endl;
}

/**
 * @brief https://atcoder.jp/contests/abc220/tasks/abc220_f
 */
void solve_abc220_f()
{
    size_t N;
    cin >> N;
    tree tree(N);
    for (size_t i = 0; i < N - 1; i++)
    {
        size_t u, v;
        cin >> u >> v;
        tree.connect(u - 1, v - 1);
    }
    using E = pair<int32_t, int64_t>;
    using V = pair<int32_t, int64_t>;
    constexpr auto add_e = [](const V &v, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t e) -> E
    {
        const auto &[s, d] = v;
        return {s, d + s};
    };
    constexpr auto merge = [](const E &l, const E &r) -> E
    {
        const auto &[ls, ld] = l;
        const auto &[rs, rd] = r;
        return {ls + rs, ld + rd};
    };
    constexpr auto add_v = [](const E &e, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t parent) -> V
    {
        const auto &[s, d] = e;
        return {s + 1, d};
    };
    constexpr auto ee = []() -> E
    { return {0, 0}; };

    const auto ans = tree.collect_for_all<E, V, add_e, merge, add_v, ee>();
    for (size_t i = 0; i < N; i++)
        cout << ans.at(i).second << endl;
}

/**
 * @brief https://atcoder.jp/contests/abc222/tasks/abc222_f
 */
void solve_abc222_f()
{
    size_t N;
    cin >> N;
    tree tree(N);
    vector<int64_t> C(N - 1);
    for (size_t i = 0; i < N - 1; i++)
    {
        size_t a, b;
        int64_t c;
        cin >> a >> b >> c;
        tree.connect(a - 1, b - 1, i);
        C.at(i) = c;
    }
    vector<int64_t> D(N);
    for (size_t i = 0; i < N; i++)
    {
        int64_t d;
        cin >> d;
        D.at(i) = d;
    }
    using E = int64_t;
    using V = int64_t;
    auto add_e = [&C](const V &v, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t e) -> E
    {
        return v + C.at(e);
    };
    auto merge = [](const E &l, const E &r) -> E
    {
        return max(l, r);
    };
    auto add_v = [&D](const E &e, [[maybe_unused]] const size_t i, [[maybe_unused]] const size_t parent) -> V
    {
        return i == parent ? e : max(e, D.at(i));
    };
    constexpr auto e = []() -> E
    { return 0; };

    const auto ans = tree.collect_for_all<E, V>(add_e, merge, add_v, e);
    for (size_t i = 0; i < N; i++)
        cout << ans.at(i) << endl;
}

int main()
{
    // solve_dp_v();
    // solve_abc220_f();
    // solve_abc222_f();
}

File: py/reroot.py
# https://github.com/cexen/procon-cexen/blob/main/py/tree.py
from collections.abc import Callable
from typing import TypeAlias, TypeVar

E_ = TypeVar("E_")
V_ = TypeVar("V_")


class TreeReroot:
    """
    v1.6 cexen.
    See `solve_*()` below for examples.
    """

    def __init__(self, n: int):
        self.n = n
        self.root = -1
        self.idxs = [-1] * n
        self.adjs = [list[tuple[int, int]]() for _ in range(n)]
        self.parents: list[tuple[int, int]] = [(-1, -1)] * n

    def connect(self, u: int, v: int, e: int = -1) -> None:
        """
        Specify e (edge index) if you need it when add_e.
        """

        assert 0 <= u < self.n
        assert 0 <= v < self.n
        self.adjs[u].append((v, e))
        self.adjs[v].append((u, e))

    def sort(self, root: int = 0) -> None:
        assert 0 <= root < self.n
        if root == self.root:
            return
        self.root = root
        self.idxs.clear()
        q = [root]
        visited = [False] * self.n
        visited[root] = True
        while q:
            i = q.pop()
            self.idxs.append(i)
            for adj in self.adjs[i]:
                j, e = adj
                if visited[j]:
                    self.parents[i] = adj
                    continue
                visited[j] = True
                q.append(j)

    def collect_for_root(
        self,
        add_e: Callable[[V_, int, int], E_],
        merge: Callable[[E_, E_], E_],
        add_v: Callable[[E_, int, int], V_],
        ee: Callable[[], E_],
        root: int = 0,
    ) -> tuple[list[V_], list[list[E_ | None]]]:
        self.sort(root)
        ans0: list[V_] = [None] * self.n  # type: ignore
        edgeses = [list[E_ | None]() for _ in range(self.n)]
        for i in reversed(self.idxs):
            parent = self.parents[i]
            edges = edgeses[i]
            accum_l = ee()
            for adj in self.adjs[i]:
                j, e = adj
                edge = None if adj == parent else add_e(ans0[j], j, e)
                edges.append(edge)
                if edge is not None:
                    accum_l = merge(accum_l, edge)
            ans0[i] = add_v(accum_l, i, parent[0])
        return ans0, edgeses

    def collect_for_all(
        self,
        add_e: Callable[[V_, int, int], E_],
        merge: Callable[[E_, E_], E_],
        add_v: Callable[[E_, int, int], V_],
        ee: Callable[[], E_],
        root: int = 0,
    ) -> list[V_]:
        _, edgeses = self.collect_for_root(add_e, merge, add_v, ee, root)
        ans: list[V_] = [None] * self.n  # type: ignore
        q = [(root, ee())]
        while q:
            i, edge_parent = q.pop()
            adjsi = self.adjs[i]
            parent = self.parents[i]
            edges = edgeses[i]
            accums_l = [ee()]
            for k, edge in enumerate(edges):
                if edge is None:
                    edge = edge_parent
                    edges[k] = edge_parent
                accums_l.append(merge(accums_l[-1], edge))
            accums_l.pop()  # not used
            accum_r = ee()
            for adj, accum_l, edge in zip(
                reversed(adjsi), reversed(accums_l), reversed(edges)
            ):
                assert edge is not None
                if adj != parent:
                    j, e = adj
                    # myself as a child of j
                    myself = add_v(merge(accum_l, accum_r), i, j)
                    q.append((j, add_e(myself, i, e)))
                accum_r = merge(edge, accum_r)
            ans[i] = add_v(accum_r, i, i)
        return ans


############
# template #
############
# E: TypeAlias = int
# V: TypeAlias = int
# def add_e(v: V, i: int, e: int) -> E:
#     raise NotImplementedError
# def merge(l: E, r: E) -> E:
#     raise NotImplementedError
# def add_v(e: E, i: int, parent: int) -> V:
#     raise NotImplementedError
# def ee() -> E:
#     raise NotImplementedError


# --------------------


def solve_dp_v():
    """
    https://atcoder.jp/contests/dp/tasks/dp_v
    """
    N, M = map(int, input().split())
    tree = TreeReroot(N)
    for _ in range(N - 1):
        x, y = map(int, input().split())
        tree.connect(x - 1, y - 1)
    E = int
    V = int

    def add_e(v: V, i: int, e: int) -> E:
        return (v + 1) % M

    def merge(l: E, r: E) -> E:
        return l * r % M

    def add_v(e: E, i: int, parent: int) -> V:
        return e

    def ee() -> E:
        return 1

    for ans in tree.collect_for_all(add_e, merge, add_v, ee):
        print(ans)


def solve_abc220_f():
    """
    https://atcoder.jp/contests/abc220/tasks/abc220_f
    """
    N = int(input())
    tree = TreeReroot(N)
    for _ in range(N - 1):
        u, v = map(int, input().split())
        tree.connect(u - 1, v - 1)

    E = tuple[int, int]
    V = tuple[int, int]

    def add_e(v: V, i: int, e: int) -> E:
        n, d = v
        return n, d + n

    def merge(l: E, r: E) -> E:
        nl, dl = l
        nr, dr = r
        return nl + nr, dl + dr

    def add_v(e: E, i: int, parent: int) -> V:
        n, d = e
        return n + 1, d

    def ee() -> E:
        return 0, 0

    for _, ans in tree.collect_for_all(add_e, merge, add_v, ee):
        print(ans)


def solve_abc222_f():
    """
    https://atcoder.jp/contests/abc222/tasks/abc222_f
    """
    N = int(input())
    tree = TreeReroot(N)
    C = [-1] * (N - 1)
    for i in range(N - 1):
        a, b, c = map(int, input().split())
        tree.connect(a - 1, b - 1, i)
        C[i] = c
    D = [int(v) for v in input().split()]

    E = int
    V = int

    def add_e(v: V, i: int, e: int) -> E:
        return v + C[e]

    def merge(l: E, r: E) -> E:
        return max(l, r)

    def add_v(e: E, i: int, parent: int) -> V:
        return max(e, D[i]) if i != parent else e

    def ee() -> E:
        return 0

    for ans in tree.collect_for_all(add_e, merge, add_v, ee):
        print(ans)
